tailnet/configmaps.go

```diff
@@ -0,0 +1,274 @@
+			PrivateKey: nodeKey,
+			PacketFilter: []filter.Match{{
+				// Allow any protocol!
+				IPProto: []ipproto.Proto{ipproto.TCP, ipproto.UDP, ipproto.ICMPv4, ipproto.ICMPv6, ipproto.SCTP},
+				// Allow traffic sourced from anywhere.
+				Srcs: []netip.Prefix{
+					netip.PrefixFrom(netip.AddrFrom4([4]byte{}), 0),
+					netip.PrefixFrom(netip.AddrFrom16([16]byte{}), 0),
+				},
+				// Allow traffic to route anywhere.
+				Dsts: []filter.NetPortRange{
+					{
+						Net: netip.PrefixFrom(netip.AddrFrom4([4]byte{}), 0),
+						Ports: filter.PortRange{
+							First: 0,
+							Last:  65535,
+						},
+					},
+					{
+						Net: netip.PrefixFrom(netip.AddrFrom16([16]byte{}), 0),
+						Ports: filter.PortRange{
+							First: 0,
+							Last:  65535,
+						},
+					},
+				},
+				Caps: []filter.CapMatch{},
+			}},
+		},
+		peers:     make(map[uuid.UUID]*peerLifecycle),
+		addresses: addresses,
+	}
+	go c.configLoop()
+	return c
+}
+
+func (c *configMaps) configLoop() {
+	c.L.Lock()
+	defer c.L.Unlock()
+	defer func() {
+		c.phase = closed
+		c.Broadcast()
+	}()
+	for {
+		for !(c.closing || c.netmapDirty || c.filterDirty || c.derpMapDirty) {
+			c.phase = idle
+			c.Wait()
+		}
+		if c.closing {
+			return
+		}
+		// queue up the reconfiguration actions we will take while we have
+		// the configMaps locked. We will execute them while unlocked to avoid
+		// blocking during reconfig.
+		actions := make([]func(), 0, 3)
+		if c.derpMapDirty {
+			derpMap := c.derpMapLocked()
+			actions = append(actions, func() {
+				c.engine.SetDERPMap(derpMap)
+			})
+		}
+		if c.netmapDirty {
+			nm := c.netMapLocked()
+			actions = append(actions, func() {
+				c.engine.SetNetworkMap(nm)
+				c.reconfig(nm)
+			})
+		}
+		if c.filterDirty {
+			f := c.filterLocked()
+			actions = append(actions, func() {
+				c.engine.SetFilter(f)
+			})
+		}
+
+		c.netmapDirty = false
+		c.filterDirty = false
+		c.derpMapDirty = false
+		c.phase = configuring
+		c.Broadcast()
+		func() {
+			// this may look a little odd, but here we want this code to run
+			// without the lock, and then relock when done
+			c.L.Unlock()
+			defer c.L.Lock()
+			for _, a := range actions {
+				a()
+			}
+		}()
+	}
+}
+
+func (c *configMaps) close() {
+	c.L.Lock()
+	defer c.L.Unlock()
+	c.closing = true
+	c.Broadcast()
+	for c.phase != closed {
+		c.Wait()
+	}
+}
+
+func (c *configMaps) netMapLocked() *netmap.NetworkMap {
+	nm := new(netmap.NetworkMap)
+	*nm = c.static
+
+	nm.Addresses = make([]netip.Prefix, len(c.addresses))
+	copy(nm.Addresses, c.addresses)
+
+	nm.DERPMap = DERPMapFromProto(c.derpMap)
+	nm.Peers = c.peerConfigLocked()
+	nm.SelfNode.Addresses = nm.Addresses
+	nm.SelfNode.AllowedIPs = nm.Addresses
+	return nm
+}
+
+func (c *configMaps) peerConfigLocked() []*tailcfg.Node {
+	out := make([]*tailcfg.Node, 0, len(c.peers))
+	for _, p := range c.peers {
+		out = append(out, p.node.Clone())
+	}
+	return out
+}
+
+func (c *configMaps) setAddresses(ips []netip.Prefix) {
```

Comment:
Even though these won't be used outside of tailnet, I think it would be clearer to export the functions that callers are expected to use. Otherwise they feel a little lost next to all of the internal-only functions.

Reply:
Do you think it's worth putting into its own package?  Otherwise I don't think the internal/external distinction is meaningful and I'm not sure readers would will pick up on the difference.

Maybe adding some method comments would help make it clear.

Reply:
As I was going through, other than the configLoop() routine, I don't know if any internal/external distinction is going to pass the test of time.  For example, right now the functions that return the current config are only called from the configLoop(), but there isn't any strong reason we couldn't call them from the Conn later---there are a lot of functions already that query various aspects of config---so long as they obey the locking requirements.

Reply:
That makes sense. I was initially a bit confused on what was intended to be called externally. Comments sound helpful 👍